src/lib/scan-imports.ts

```diff
@@ -0,0 +1,160 @@
+import fs from 'fs';
+import chalk from 'chalk';
+import ora from 'ora';
+import glob from 'glob';
+import {parse} from '@babel/parser';
+import traverse, {NodePath} from '@babel/traverse';
+import {
+  CallExpression,
+  ImportDeclaration,
+  isImportDeclaration,
+  isCallExpression,
+} from '@babel/types';
+
+/**
+ * Splits a filename into folder segments
+ */
+function explode(file: string) {
+  return file.split('/').filter(part => part);
+}
+
+/**
+ * Attempts to figure out a module’s npm name
+ */
+function getNpmName(modulePath: string, dest: string) {
+  // if local module, ignore
+  if (modulePath.startsWith('.')) {
+    return modulePath;
+  }
+
+  // if there’s no slash, this is likely a special resolution case and should be ignored
+  if (!modulePath.includes('/')) {
+    return modulePath;
+  }
+
+  let moduleName = modulePath;
+  const lastPart = [...explode(dest)].pop();
+
+  // this is likely a web_module, based on --dest
+  if (moduleName.includes(lastPart)) {
+    const [_, webModule] = moduleName.split(lastPart);
+    if (webModule) {
+      moduleName = webModule.replace(/\.[^.]+$/, ''); // remove trailing extension
+    }
+  }
+
+  // only return top-level package name if there are slashes
+  const [scope, pkg] = explode(moduleName);
+  return scope.startsWith('@') ? `${scope}/${pkg}` : scope;
+}
+
+/**
+ * Return import name (or undefined)
+ */
+function importModuleName(node: NodePath<ImportDeclaration>): string | undefined {
+  if (isImportDeclaration(node)) {
+    return node.node.source.value;
+  }
+  return undefined;
+}
+
+/**
+ * Return dynamic import name (or undefined)
+ */
+function dynamicImportModuleName(node: NodePath<CallExpression>): string | undefined {
+  if (isCallExpression(node)) {
+    if (node.node.callee.type === 'Import' && node.node.arguments && node.node.arguments.length) {
+      return (node.node.arguments[0] as any).value;
+    }
+  }
+  return undefined;
+}
+
+/**
+ * Return array of dependencies
+ */
+function parseImports(file: string) {
+  const code = fs.readFileSync(file, 'utf8');
+
+  // skip file if empty, or if permissions error
+  if (!code) {
+    return [];
+  }
+  const deps = new Set<string>(); // micro-optimization: use Set() to dedupe
+
+  // try to parse; this will fail if not JS
+  try {
```

Comment:
When switching to recursive glob, I noticed it failed if you threw non-JS files at it (or JS files with a syntax error).

This is a bit safer now—it either parse, or it don’t. And it keeps moving along 💪 (but should we raise a warning? My feeling is that’d be a bit annoying).

Reply:
Definitely no need to error if it's non-JS. it could make sense to error on a JS file that can't be parsed thought. Up to you!

Reply:
Cool. A bit more context: I’ve worked with MIME types and extensions, and I went with a “try to parse it and see what happens” route because MIME isn’t guaranteed in many cases. File extensions just seemed like extra management when someone invariably has a different setup.

In my experience, though, the Babel parser seems to fail very quickly. IMO this is the most stable solution that caters to the widest range of setups, without management on our part or performance degradation. Anyway, all just extra thinking over this course of action.

Reply:
Ah so does it mean TypeScript is not supported?